### PR TITLE
Fix zombie task caused by early task failure

### PR DESCRIPTION
**Pull request by xiaoxmeng**
In Prestissimo, a velox task can be terminated by coordinator even before
it starts execution. This could lead to zombie task as the output buffer manager
will hold a reference on the task, and only the task terminate will remove the
task from the output buffer manager.
This PR checks the early task termination during the task start and verified with
unit test.

### CODE REVIEW

File: velox/exec/Task.cpp [hunk 1]
```diff
@@ -614,52 +635,62 @@ void Task::start(
...
-    // after sizing 'drivers_' but before starting the
-    // Drivers. 'drivers_' can be read by memory recovery or
-    // cancellation while Drivers are being made, so the array should
-    // have final size from the start.
-    auto bufferManager = self->bufferManager_.lock();
-    VELOX_CHECK_NOT_NULL(
-        bufferManager,
-        "Unable to initialize task. "
-        "PartitionedOutputBufferManager was already destructed");
+      // Register self for possible memory recovery callback. Do this
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:639]: Not your code, but I don't see what this comment refers to.

> so the array should have final size from the start.

Which array is this about?

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 2]
```diff
@@ -1780,5 +1780,49 @@ DEBUG_ONLY_TEST_F(MultiFragmentTest, exchangeStatsOnFailure) {
...
+  auto finalSortPlan = PlanBuilder()
+                           .mergeExchange(outputType, {"c0"})
+                           .partitionedOutput({}, 1)
+                           .planNode();
+
+  auto finalSortTask = makeTask(finalSortTaskId, finalSortPlan, 0);
+  try {
+    VELOX_FAIL("memoryAbortTest");
+  } catch (const VeloxRuntimeError& e) {
+    finalSortTask->pool()->abort(std::current_exception());
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:1815]: Which code path in Prestissimo ends up calling task->pool()->abort without calling task->requestAbort first?

File: velox/exec/Task.cpp [hunk 3]
```diff
@@ -553,6 +553,20 @@ void Task::start(
   facebook::velox::process::ThreadDebugInfo threadDebugInfo{
       self->queryCtx()->queryId(), self->taskId_, nullptr};
   facebook::velox::process::ScopedThreadDebugInfo scopedInfo(threadDebugInfo);
+
+#define CHECK_EARLY_TASK_ERROR()                                            \
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:557]: Macros are hard to read. What's the motivation for using a macro here instead of a regular function / method?

File: velox/exec/Task.cpp [hunk 4]
```diff
@@ -564,8 +578,15 @@ void Task::start(
...
 
     uint32_t numPipelines;
+    std::shared_ptr<const core::PartitionedOutputNode> partitionedOutputNode{
+        nullptr};
+    int numOutputDrivers{0};
     {
       std::unique_lock<std::mutex> l(self->mutex_);
+      // NOTE: the driver might have been terminated by async external event
+      // even before it gets started.
+      CHECK_EARLY_TASK_ERROR();
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:588]: Not your code, but this method grew very large and hard to follow over time. It would be nice to refactor it first before introducing more logic.

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Thank you for investigating zombie tasks. Some questions.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/tests/MultiFragmentTest.cpp:1815]: This is triggered by memory arbitration internally when memory arbitration fails to allocate memory and has to abort one. I just run into this when I add memory pool abort test (which is supposed not happening in production) and found the task hanging problem. And realized that similar race could happen in Prestissimo with the concurrent updates from coordinator.

File: velox/exec/Task.cpp (hunk 3 above)
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:557]: You are right. It seems that a function is more appropriate here. Will change.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/exec/tests/MultiFragmentTest.cpp:1815]: > And realized that similar race could happen in Prestissimo with the concurrent updates from coordinator.

I'm not sure I understand. Which API call from coordinator would end up calling pool->abort and not task->requestAbort or some other Task API?

File: velox/exec/Task.cpp [hunk 5]
```diff
@@ -543,13 +543,19 @@ RowVectorPtr Task::next(ContinueFuture* future) {
   }
 }
 
+#ifdef VELOX_ENABLE_BACKWARD_COMPATIBILITY
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:546]: This macro doesn't work in .cpp, only in headers.

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 6]
```diff
@@ -1783,5 +1780,61 @@ DEBUG_ONLY_TEST_F(MultiFragmentTest, exchangeStatsOnFailure) {
   ASSERT_TRUE(waitForTaskCompletion(producerTask.get(), 3'000'000));
 }
 
+TEST_F(MultiFragmentTest, earlyTaskFailure) {
+  setupSources(1, 10);
+  std::vector<std::shared_ptr<TempFilePath>> filePaths(
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:1785]: Why not auto filePaths = filePaths_? Why do we need this copy?

File: velox/exec/tests/MultiFragmentTest.cpp [hunk 7]
```diff
@@ -1783,5 +1780,61 @@ DEBUG_ONLY_TEST_F(MultiFragmentTest, exchangeStatsOnFailure) {
...
+                                      .planNode()})
+                             .partitionedOutput({}, 1)
+                             .planNode();
+  for (bool internalFailure : {false, true}) {
+    SCOPED_TRACE(fmt::format("internalFailure: {}", internalFailure));
+
+    auto partialSortTask = makeTask(partialSortTaskId, partialSortPlan, 1);
+    partialSortTask->start(1);
+    addHiveSplits(partialSortTask, filePaths);
+    RowTypePtr outputType = partialSortPlan->outputType();
```
Comment (mbasmanova) [inline velox/exec/tests/MultiFragmentTest.cpp:1805]: auto

**Review by mbasmanova — COMMENTED** [review]
Thanks.

File: velox/exec/tests/MultiFragmentTest.cpp (hunk 6 above)
Comment (xiaoxmeng) [reply velox/exec/tests/MultiFragmentTest.cpp:1785]: We don't need it. I copy over from the other test. Thanks!

**Review by mbasmanova — APPROVED** [review]
Thanks.
